### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by iomodo**
This is the initial implementation of the Mattermost plugin suggestions

### CODE REVIEW

**Comment by lieut-data** [comment]
I've actually reviewed this once in the past, so thinking it makes sense for @jfrerich to chime in. Let me know, @iomodo, if there's anything specific you'd want me to take a look at!

**Comment by jfrerich** [comment]
Thanks @lieut-data for the review ask.  This is a large PR, but a quick glance I notice it's server-side so I'm comfortable with the task.  @iomodo, since I will be a fresh set of eyes on this code, please let me know if there is an area you would special attention.  I'll work this in in the next few days!

File: public/hello.html [hunk 1]
```diff
@@ -0,0 +1 @@
+Hello from the static files public folder for the com.mattermost.plugin-starter-template plugin!
```
Comment (jfrerich) [inline public/hello.html]: This dir isn't actually used.  Perhaps delete from the repo?  For reference, jira, github, and zoom plugins have all removed this directory.

File: server/command.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	addRandomChannelAction = "add"
+	resetAction            = "reset"
+	computeAction          = "compute"
+
+	displayName          = "Suggestions"
+	desc                 = "Mattermost Suggestions Plugin"
+	noNewChannelsText    = "No new channels for you."
+	addRandomChannelText = "Channel was successfully added."
+	resetText            = "Recommendations were cleared."
+	computeText          = "Recomendations were computed."
```
Comment (jfrerich) [inline server/command.go]: s/Recomendations/Recommendations/

File: server/api_helper.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+			}
+			teamUsers[team.Id] = append(teamUsers[team.Id], users...)
+			page++
+		}
+	}
+	return teamUsers, nil
+}
+
+// RunOnSingleNode will run function f on only a single node in a HA environment
+func (p *Plugin) RunOnSingleNode(f func()) error {
```
Comment (iomodo) [inline server/api_helper.go:83]: @lieut-data, @jfrerich 
Code is mainly reviewed. I have some concerns about this method. If you could review it

File: README.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# Mattermost Suggestions Plugin
+[![CircleCI](https://circleci.com/gh/mattermost/mattermost-plugin-suggestions.svg?style=svg)](https://circleci.com/gh/mattermost/mattermost-plugin-suggestions)
+[![codecov](https://codecov.io/gh/mattermost/mattermost-plugin-suggestions/branch/master/graph/badge.svg)](https://codecov.io/gh/mattermost/mattermost-plugin-suggestions)
+
+This plugin delivers a channel suggestions for the users using [collaborative filtering](https://en.wikipedia.org/wiki/Collaborative_filtering).
```
Comment (jfrerich) [inline README.md:3]: s/a channel/ channel/

File: test.sh [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
```
Comment (jfrerich) [inline test.sh]: Is this file used?  Perhaps it should be included in a Make target so developers won't need to remember to run manually, otherwise removed?

File: server/command.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+* |/suggest compute| - Computes suggestions. For testing only.
+`
+
+func getCommand() *model.Command {
+	return &model.Command{
+		Trigger:          trigger,
+		DisplayName:      displayName,
+		Description:      desc,
+		AutoComplete:     true,
+		AutoCompleteDesc: "Available commands: channels, help",
```
Comment (jfrerich) [inline server/command.go:36]: Should we include `reset` and `compute` options in the AutoCompleteDesc?

File: server/command.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+	if appErr != nil {
+		return nil, appErr
+	}
+	if len(channels) == 0 {
+		p.postCommandResponse(args, noNewChannelsText)
+		return &model.CommandResponse{}, nil
+	}
+	text := "Channels we recommend\n"
+	for _, channel := range channels {
+		text += " * ~" + channel.Name + " - " + channel.Purpose + "\n"
```
Comment (jfrerich) [inline server/command.go]: Sometimes `channel.Purpose` is `nil`.  Propose adding this check so the resulting text doesn't end up with the `channel.Name` followed only by a dash.

Another addition that would be helpful is to add get the `channel.TeamId` and corresponding `TeamName` so the user knows which team the channel exists.  If we decide to add this extra information, we need to make sure the user only gets suggestions for teams they have permissions to view.

**Review by jfrerich — COMMENTED** [review]
Thanks, @iomodo and this is awesome work!  I have a few non-blocking comments for discussion.  If we agree some are needed, but would take too long to address, we can cut a few Issues in github and get this thing merged first!

File: server/api_helper.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+	ok, err := p.Helpers.KVCompareAndSetJSON("RunOnSingleNode", nil, 0)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		f()
+	}
+	p.Helpers.KVCompareAndSetJSON("RunOnSingleNode", 0, nil) //TODO what to do if not saved?
```
Comment (lieut-data) [inline server/api_helper.go]: As a short term workaround to having automatic support for the expiration, what if instead of setting the value `0`, we set the current timestamp, and if we fail to acquire, check the timestamp and if it's greater than some threshold,  try to `KVCompareAndSetJSON` using that value instead?

File: server/command.go (hunk 6 above)
Comment (iomodo) [reply server/command.go:36]: `reset` and `compute` were used for testing only. They should be removed soon

File: server/api_helper.go (hunk 8 above)
Comment (iomodo) [reply server/api_helper.go]: It's a distributed consensus problem, so there is no easy solution. This workaround will increase complexity but will have the same flaw - `KVCompareAndSetJSON` not working. And in addition we will be relying on the timestamp working right(synchronized) on every node. Probably better to use `KVCompareAndSetWithExpiryJSON`?

For this concrete plugin current implementation is not a big problem, because if `KVCompareAndSetJSON` will not delete the key on the first time cron job will try to do it again.
I'm more concerned about the `RunOnSingleNode` implementation as a standalone function.

Comment (lieut-data) [reply server/api_helper.go]: @iomodo, can you elaborate on this being a distributed consensus problem? `CompareAndSet` avoids the distributed issues by relying on an atomic write to the database. It is impossible for more than one node to race with another node using `CompareAndSet` and believe they both write a /new/ value.

I think we can rely on a reasonable degree of synchronization with respect to the timestamps between servers, but I note that the correctness of the atomicity doesn't rely on this, only the decision to forcibly expire the lock previously acquired. If the timestamps are off by more than a few seconds, posts from one server will appear out of sync with posts from another, and we have a larger problem at the application level.

`KVCompareAndSetWithExpiryJSON` doesn't help here, since it's just literally the above algorithm embedded into the server. It probably could (should?) have been a plugin helper instead.

File: .circleci/config.yml [hunk 9]
```diff
@@ -0,0 +1,36 @@
+version: 2.1
+executors:
+  default:
+    docker:
+      - image: circleci/golang:1.12-node
+
+jobs:
```
Comment (hanzei) [inline .circleci/config.yml:7]: Why are you not using the default jobs from the circleci orb like in https://github.com/mattermost/mattermost-plugin-jira/blob/master/.circleci/config.yml?

File: README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+## Usage
+Trigger the suggestion via the slash command `/suggest channels`. Other triggers may be added later.
+
+## Future Work
+* Change user activity score and add more features.
+* Implement other machine learning models
+* Collect user data, perform tests and validation, optimize parameters, improve Root Mean Square Error
+
+#### Nice to have methods in Mattermost Plugin API
```
Comment (hanzei) [inline README.md]: 2/5 that this should not be part of the readme

File: go.mod [hunk 11]
```diff
@@ -0,0 +1,18 @@
+module github.com/mattermost/mattermost-plugin-suggestions
+
+go 1.12
+
+require (
+	github.com/blang/semver v3.6.1+incompatible // indirect
+	//github.com/mattermost/mattermost-server v5.12.0+incompatible
```
Comment (hanzei) [inline go.mod]: Please remove this

File: go.mod [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+	github.com/mattermost/mattermost-server v1.4.1-0.20190911153151-98489b9e67d9
+	github.com/pkg/errors v0.8.1
+	github.com/robfig/cron v1.2.0
+	github.com/stretchr/testify v1.3.0
+	willnorris.com/go/imageproxy v0.9.0
```
Comment (hanzei) [inline go.mod]: Why is this required?

File: go.mod [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+go 1.12
+
+require (
+	github.com/blang/semver v3.6.1+incompatible // indirect
+	//github.com/mattermost/mattermost-server v5.12.0+incompatible
+	github.com/mattermost/mattermost-server v1.4.1-0.20190911153151-98489b9e67d9
+	github.com/pkg/errors v0.8.1
+	github.com/robfig/cron v1.2.0
+	github.com/stretchr/testify v1.3.0
```
Comment (hanzei) [inline go.mod]: Nit:Please update to 1.4.0

File: go.mod [hunk 14]
```diff
@@ -0,0 +1,18 @@
+module github.com/mattermost/mattermost-plugin-suggestions
+
+go 1.12
+
+require (
+	github.com/blang/semver v3.6.1+incompatible // indirect
+	//github.com/mattermost/mattermost-server v5.12.0+incompatible
+	github.com/mattermost/mattermost-server v1.4.1-0.20190911153151-98489b9e67d9
+	github.com/pkg/errors v0.8.1
+	github.com/robfig/cron v1.2.0
```
Comment (hanzei) [inline go.mod]: Nit: Is there a specific reason you are not using the latest version?

File: go.mod [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+	github.com/stretchr/testify v1.3.0
+	willnorris.com/go/imageproxy v0.9.0
+)
+
+// Workaround for https://github.com/golang/go/issues/30831 and fallout.
+//replace github.com/golang/lint => github.com/golang/lint v0.0.0-20190227174305-8f45f776aaf1
```
Comment (hanzei) [inline go.mod:14]: Why is this commented out?

File: server/api_helper.go [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+import (
+	"github.com/mattermost/mattermost-server/model"
+)
+
+// GetAllUsers returns all users
+func (p *Plugin) GetAllUsers() (map[string]*model.User, *model.AppError) {
+	allUsers := make(map[string]*model.User)
+	teamUsers, err := p.getTeamUsers()
+	if err != nil {
+		return nil, err
```
Comment (hanzei) [inline server/api_helper.go]: Nit: Wrap the error here and return `error` instead of `*model.AppError`.

File: server/api_helper.go [hunk 17]
```diff
@@ -0,0 +1,105 @@
+package main
+
+import (
+	"github.com/mattermost/mattermost-server/model"
+)
+
+// GetAllUsers returns all users
+func (p *Plugin) GetAllUsers() (map[string]*model.User, *model.AppError) {
```
Comment (hanzei) [inline server/api_helper.go]: You are never using the method. Why is it here?

File: server/api_helper.go [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+	for _, users := range teamUsers {
+		for _, user := range users {
+			allUsers[user.Id] = user
+		}
+	}
+	return allUsers, nil
+}
+
+// GetAllChannels returns all channels
+func (p *Plugin) GetAllChannels() (map[string]*model.Channel, *model.AppError) {
```
Comment (hanzei) [inline server/api_helper.go]: You are only using the channelId. I think you should either return a list of channelIds or a list of channels.

File: server/api_helper.go [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+			for _, channel := range channels {
+				allChannels[channel.Id] = channel
+			}
+		}
+	}
+	return allChannels, nil
+}
+
+// GetAllPublicChannelsForUser returns all public channels for user
+func (p *Plugin) GetAllPublicChannelsForUser(userID string) (map[string]*model.Channel, *model.AppError) {
```
Comment (hanzei) [inline server/api_helper.go]: Same comment here.

File: server/api_helper.go [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+			teamUsers[team.Id] = append(teamUsers[team.Id], users...)
+			page++
+		}
+	}
+	return teamUsers, nil
+}
+
+// RunOnSingleNode will run function f on only a single node in a HA environment
+func (p *Plugin) RunOnSingleNode(f func()) error {
+	ok, err := p.Helpers.KVCompareAndSetJSON("RunOnSingleNode", nil, 0)
```
Comment (hanzei) [inline server/api_helper.go]: If I call `RunOnSingleNode` with two different functions only one of the functions will be executed, right?

File: server/channel_recommendations.go [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+import (
+	"sort"
+
+	"github.com/mattermost/mattermost-plugin-suggestions/server/ml"
+	"github.com/mattermost/mattermost-server/mlog"
+	"github.com/mattermost/mattermost-server/model"
+)
+
+const numberOfRecommendedChannels = 5
+const numberOfNeighbors = 10 // TODO exact number of nearest neighbors should be determined from real data
```
Comment (hanzei) [inline server/channel_recommendations.go:11]: 1/5 that this should be an GH issue instead of a comment

File: server/channel_recommendations.go [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+
+func (p *Plugin) getChannelListFromRecommendations(recommendations []*recommendedChannel) ([]*model.Channel, *model.AppError) {
+	sort.Slice(recommendations, func(i, j int) bool {
+		return recommendations[i].Score > recommendations[j].Score
+	})
+	n := min(numberOfRecommendedChannels, len(recommendations))
+	channels := make([]*model.Channel, 0, n)
+	for i := 0; i < n; i++ {
+		channel, err := p.API.GetChannel(recommendations[i].ChannelID)
+		println(err)
```
Comment (hanzei) [inline server/channel_recommendations.go]: Debug code?

File: server/channel_recommendations.go [hunk 23]
```diff
@@ -0,0 +1,84 @@
...
+		if err != nil {
+			return nil, err
+		}
+		channels = append(channels, channel)
+	}
+	return channels, nil
+}
+
+func (p *Plugin) preCalculateRecommendations() *model.AppError {
```
Comment (hanzei) [inline server/channel_recommendations.go:44]: 2/5 that you return an `error` here.

File: server/command.go [hunk 24]
```diff
@@ -0,0 +1,135 @@
...
+	displayName          = "Suggestions"
+	desc                 = "Mattermost Suggestions Plugin"
+	noNewChannelsText    = "No new channels for you."
+	addRandomChannelText = "Channel was successfully added."
+	resetText            = "Recommendations were cleared."
+	computeText          = "Recommendations were computed."
+)
+
+const commandHelp = `
+* |/suggest channels| - Suggests relevant channels for the user
```
Comment (hanzei) [inline server/command.go:25]: Nit: You can build the string with `"` and `+`. This way you don't need `helpResponse()`.

File: server/ml/priority_queue.go [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+
+// An Item is something we manage in a max heap.
+type Item struct {
+	value    interface{}
+	priority float64
+}
+
+// A MaxHeapK implements heap.Interface and holds k Items with highest priority
+type MaxHeapK struct {
+	elem []*Item
```
Comment (hanzei) [inline server/ml/priority_queue.go]: Nit:
```suggestion
	elements []*Item
```

File: server/ml/priority_queue.go [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+	"container/heap"
+)
+
+// An Item is something we manage in a max heap.
+type Item struct {
+	value    interface{}
+	priority float64
+}
+
+// A MaxHeapK implements heap.Interface and holds k Items with highest priority
```
Comment (hanzei) [inline server/ml/priority_queue.go]: Nit: `k` is not public and hence not important
```suggestion
// A MaxHeapK implements heap.Interface.
```

File: server/ml/priority_queue.go [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+type MaxHeapK struct {
+	elem []*Item
+	k    int
+}
+
+// NewMaxHeapK returns a brand new max heap with capacity k
```
Comment (hanzei) [inline server/ml/priority_queue.go]:
```suggestion
// NewMaxHeapK returns a brand new max heap with capacity k.
```

File: server/ml/priority_queue.go [hunk 28]
```diff
@@ -0,0 +1,57 @@
...
+func NewMaxHeapK(k int) *MaxHeapK {
+	mh := new(MaxHeapK)
+	mh.k = k
+	mh.elem = make([]*Item, 0, k)
+	return mh
+}
+
+func (pq MaxHeapK) Len() int { return len(pq.elem) }
```
Comment (hanzei) [inline server/ml/priority_queue.go]: Nit: add a godoc comment

File: server/ml/priority_queue.go [hunk 29]
```diff
@@ -0,0 +1,57 @@
...
+
+func (pq MaxHeapK) Less(i, j int) bool {
```
Comment (hanzei) [inline server/ml/priority_queue.go:31]: Same

File: server/ml/priority_queue.go [hunk 30]
```diff
@@ -0,0 +1,57 @@
...
+	return pq.elem[i].priority < pq.elem[j].priority
+}
+
+func (pq MaxHeapK) Swap(i, j int) {
```
Comment (hanzei) [inline server/ml/priority_queue.go:36]: Same

File: server/ml/simple_knn_estimator.go [hunk 31]
```diff
@@ -0,0 +1,147 @@
...
+
+// NewSimpleKNN returns Simple KNN Estimator
+func NewSimpleKNN(params map[string]interface{}) BaseEstimator {
+	simpleKNN := new(SimpleKNN)
+	simpleKNN.SetParams(params)
+	return simpleKNN
+}
+
+// SetParams sets parameters for KNN estimator
+func (knn *SimpleKNN) SetParams(params map[string]interface{}) {
```
Comment (hanzei) [inline server/ml/simple_knn_estimator.go:35]: Nit: You maybe want to use something like https://github.com/tmrts/go-patterns/blob/master/idiom/functional-options.md here

File: server/ml/simple_knn_estimator.go [hunk 32]
```diff
@@ -0,0 +1,147 @@
...
+	neighbors := knn.getNeighbors(channel)
+
+	score := 0.0
+	sum := 0.0
+	for i := 0; i < len(neighbors); i++ {
+		score += knn.channelSimilarityMatrix[channel][neighbors[i]] * knn.activityMatrix[neighbors[i]][user]
+		sum += knn.channelSimilarityMatrix[channel][neighbors[i]]
+	}
+	if sum != 0 {
+		score = score / sum
```
Comment (hanzei) [inline server/ml/simple_knn_estimator.go]: Nit:
```suggestion
		score /= sum
```

File: server/plugin.go [hunk 33]
```diff
@@ -0,0 +1,115 @@
...
+
+	// a job for pre-calculating channel recommendations for users.
+	preCalcJob    *cron.Cron
+	preCalcPeriod string
+	botUserID     string
+}
+
+type readFile func(filename string) ([]byte, error)
+
+func (p *Plugin) setupBot(reader readFile) error {
```
Comment (hanzei) [inline server/plugin.go:33]: Why are you passing this as a parameter?

File: plugin.json [hunk 34]
```diff
@@ -0,0 +1,24 @@
...
+        }
+    },
+    "settings_schema": {
+        "header": "The Mattermost Suggestions Plugin recommends interesting channels to the user",
+        "footer": "",
+        "settings": [{
+            "key": "PreCalculationPeriod",
+            "display_name": "Precalculation period",
+            "type": "text",
+            "help_text": "Channel recommendations are precalculated periodically. One can choose the period based on [cron specifications](https://godoc.org/github.com/robfig/cron). Examples: \n @daily - Run once a day, midnight. \n @hourly - Run once an hour, beginning of hour. \n @every 1h30m - Run every hour and thirty minutes. \n The default is @weekly"
```
Comment (hanzei) [inline plugin.json]: I this you want to add a default value here: https://developers.mattermost.com/extend/plugins/manifest-reference/#settings_schema.settings.default

File: server/plugin.go [hunk 35]
```diff
@@ -0,0 +1,115 @@
...
+	appErr := p.API.SetProfileImage(botID, profileImage)
+	if appErr != nil {
+		return errors.Wrap(appErr, "couldn't set profile image")
+	}
+	return nil
+}
+
+func (p *Plugin) startPrecalcJob() error {
+	config := p.getConfiguration()
+	p.preCalcPeriod = "@weekly" // Run once a week, midnight between Sat/Sun
```
Comment (hanzei) [inline server/plugin.go]: 1/5 you should use the default value instead: https://developers.mattermost.com/extend/plugins/manifest-reference/#settings_schema.settings.default

File: server/channel_recommendations.go [hunk 36]
```diff
@@ -0,0 +1,84 @@
...
+	mlog.Info("preCalculateRecommendations")
```
Comment (hanzei) [inline server/channel_recommendations.go]: Please use `p.API.LogInfo` instead

File: server/plugin.go [hunk 37]
```diff
@@ -0,0 +1,115 @@
...
+	if config.PreCalculationPeriod != "" {
+		p.preCalcPeriod = config.PreCalculationPeriod
+	}
+	c := cron.New()
+	if err := c.AddFunc(p.preCalcPeriod, func() {
```
Comment (hanzei) [inline server/plugin.go]: 2/5 that `func` should be a method of `Plugin` that uses `p.API.Log*` for logging.

File: server/plugin.go [hunk 38]
```diff
@@ -0,0 +1,115 @@
...
+	p.preCalcJob = c
+	return nil
+}
+
+// OnActivate will be run on plugin activation.
+func (p *Plugin) OnActivate() error {
+	p.API.RegisterCommand(getCommand())
+	err := p.setupBot(ioutil.ReadFile)
+	if err != nil {
+		return err
```
Comment (hanzei) [inline server/plugin.go:87]: Nit: Please `Wrap` the error.

File: server/plugin_test.go [hunk 39]
```diff
@@ -0,0 +1,106 @@
...
+		helpers := &plugintest.Helpers{}
+		helpers.On("EnsureBot", mock.Anything).Return("bot", nil)
+		plugin.SetHelpers(helpers)
+		defer helpers.AssertExpectations(t)
+		api := &plugintest.API{}
+		api.On("GetBundlePath").Return("some path", nil)
+		api.On("SetProfileImage", mock.Anything, mock.Anything).Return(model.NewAppError("", "", nil, "", 404))
+		plugin.SetAPI(api)
+		defer api.AssertExpectations(t)
+		err := plugin.setupBot(mockReader)
```
Comment (hanzei) [inline server/plugin_test.go:69]: It's not worth to test this just for the code coverage IMO. `plugin.setupBot` is not a public method and hence should not be tested.

File: server/user_activity.go [hunk 40]
```diff
@@ -0,0 +1,125 @@
...
+
+	"github.com/mattermost/mattermost-server/model"
+)
+
+type userChannelActivity = map[string]map[string]int64 //map[userID]map[channelID]activity
+
+func sizeof(a userChannelActivity) int {
+	size := uintptr(0)
+	for key, val := range a {
+		size += unsafe.Sizeof(key)
```
Comment (hanzei) [inline server/user_activity.go]: Why are no not iterating through the map?

File: server/ml/simple_knn_estimator_test.go [hunk 41]
```diff
@@ -0,0 +1,226 @@
...
+	knn.computeActivityMatrix(m)
+	assert.Equal(4, len(knn.activityMatrix))
+	for i := 0; i < 4; i++ {
+		assert.Equal(2, len(knn.activityMatrix[i]))
+	}
+	for i := 0; i < 2; i++ {
+		zeros := 0
+		ones := 0
+		for j := 0; j < 4; j++ {
+			if knn.activityMatrix[j][i] == 0 {
```
Comment (hanzei) [inline server/ml/simple_knn_estimator_test.go:73]: Nit: use `switch` here

File: server/command_test.go [hunk 42]
```diff
@@ -0,0 +1,197 @@
...
+		args := &model.CommandArgs{
+			Command: "/suggest random",
+		}
+		resp, err := plugin.ExecuteCommand(nil, args)
+		assert.Nil(t, err)
+		assert.Equal(t, &model.CommandResponse{}, resp)
+	})
+
+	t.Run("suggest command, empty action", func(t *testing.T) {
+		user := "user"
```
Comment (hanzei) [inline server/command_test.go]: Nit: Move this at the start of the test and make it a constant

File: server/command_test.go [hunk 43]
```diff
@@ -0,0 +1,197 @@
...
+		channel := "channel"
```
Comment (hanzei) [inline server/command_test.go]: Same here

File: server/command.go [hunk 44]
```diff
@@ -0,0 +1,135 @@
...
+	if err != nil {
+		errorMessage = err.Error()
+	}
+	return model.NewAppError("Suggestions Plugin", message, nil, errorMessage, http.StatusBadRequest)
+}
+
+func (p *Plugin) suggestChannelResponse(args *model.CommandArgs) (*model.CommandResponse, *model.AppError) {
+	recommendations, err := p.retreiveUserRecommendations(args.UserId)
+	if err != nil {
+		return nil, appError("Can't retreive user recommendations.", err)
```
Comment (hanzei) [inline server/command.go]:
```suggestion
		return nil, appError("can't retrieve user recommendations", err)
```

File: server/command_test.go [hunk 45]
```diff
@@ -0,0 +1,197 @@
...
+		})
+		defer helpers.AssertExpectations(t)
+		api.On("GetChannel", mock.Anything).Return(nil, model.NewAppError("", "", nil, "", 404))
+		plugin.SetAPI(api)
+		defer api.AssertExpectations(t)
+		_, err := plugin.ExecuteCommand(nil, args)
+		assert.NotNil(t, err)
+	})
+
+	t.Run("retreive user recommendations error", func(t *testing.T) {
```
Comment (hanzei) [inline server/command_test.go]:
```suggestion
	t.Run("retrieve user recommendations error", func(t *testing.T) {
```

File: server/ml/simple_knn_estimator.go [hunk 46]
```diff
@@ -0,0 +1,147 @@
...
+import (
+	"container/heap"
+	"errors"
+)
+
+const defaultK = 10
+
+// SimpleKNN struct
+type SimpleKNN struct {
+	params                  map[string]interface{}
```
Comment (hanzei) [inline server/ml/simple_knn_estimator.go]: This is unused. Can it be removed?

File: server/channel_recommendations.go [hunk 47]
```diff
@@ -0,0 +1,84 @@
...
+				}
+				if score != 0 {
+					recommendedChannels = append(recommendedChannels, &recommendedChannel{
+						ChannelID: channel.Id,
+						Score:     score,
+					})
+				}
+			}
+		}
+		p.saveUserRecommendations(userID, recommendedChannels)
```
Comment (hanzei) [inline server/channel_recommendations.go]: Please check the error value here

File: server/configuration.go [hunk 48]
```diff
@@ -0,0 +1,89 @@
...
+	// Load the public configuration fields from the Mattermost server configuration.
+	if err := p.API.LoadPluginConfiguration(configuration); err != nil {
+		return errors.Wrap(err, "failed to load plugin configuration")
+	}
+
+	p.setConfiguration(configuration)
+
+	if p.preCalcJob != nil && p.preCalcPeriod != configuration.PreCalculationPeriod {
+		p.preCalcJob.Stop()
+		p.startPrecalcJob()
```
Comment (hanzei) [inline server/configuration.go]: Please check the error value

File: server/store.go [hunk 49]
```diff
@@ -0,0 +1,83 @@
+package main
+
+import "time"
+
+const (
+	timestampKey           = "timestamp"
+	userChannelActivityKey = "userChannelActivity"
```
Comment (hanzei) [inline server/store.go]: This is unused. Can it be removed?

File: server/command.go [hunk 50]
```diff
@@ -0,0 +1,135 @@
...
+	trigger                = "suggest"
+	channelAction          = "channels"
+	addRandomChannelAction = "add"
+	resetAction            = "reset"
+	computeAction          = "compute"
+
+	displayName          = "Suggestions"
+	desc                 = "Mattermost Suggestions Plugin"
+	noNewChannelsText    = "No new channels for you."
+	addRandomChannelText = "Channel was successfully added."
```
Comment (hanzei) [inline server/command.go]: This is unused. Can it be removed?

File: server/command.go [hunk 51]
```diff
@@ -0,0 +1,135 @@
...
+	"strings"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/plugin"
+)
+
+const (
+	trigger                = "suggest"
+	channelAction          = "channels"
+	addRandomChannelAction = "add"
```
Comment (hanzei) [inline server/command.go]: This is unused. Can it be removed?

File: server/user_activity.go [hunk 52]
```diff
@@ -0,0 +1,125 @@
...
+import (
+	"time"
+	"unsafe"
+
+	"github.com/mattermost/mattermost-server/model"
+)
+
+type userChannelActivity = map[string]map[string]int64 //map[userID]map[channelID]activity
+
+func sizeof(a userChannelActivity) int {
```
Comment (hanzei) [inline server/user_activity.go]: This is unused. Can it be removed?

File: server/api_helper_test.go [hunk 53]
```diff
@@ -0,0 +1,189 @@
...
+	t.Run("getTeamUsers error", func(t *testing.T) {
+		plugin, api := getErrorFuncPlugin("GetTeams")
+		defer api.AssertExpectations(t)
+		_, err := plugin.GetAllUsers()
+		assert.NotNil(t, err)
+	})
+
+	t.Run("No error", func(t *testing.T) {
+		correctUsers := map[string]*model.User{
+			"userID1": &model.User{Id: "userID1"},
```
Comment (hanzei) [inline server/api_helper_test.go]: Please format all files with `gofmt -s`.

**Review by hanzei — CHANGES_REQUESTED** [review]
Nit: You can remove `assets/.gitkeep`

I leaved a lot of nit pick comments. The big outstanding concern to be is error handling. I think you should use `error` is most of the places and wrap it, to make debugging easier.

**Review by hanzei — CHANGES_REQUESTED** [review]
Nit: You can remove `assets/.gitkeep`

I leaved a lot of nit pick comments. The big outstanding concern to be is error handling. I think you should use `error` is most of the places and wrap it, to make debugging easier.

**Comment by hanzei** [comment]
@iomodo Let me know, when this is ready for a next review

**Comment by jasonblais** [comment]
@iomodo is this one ready for a re-review?
